File: paging/app/src/main/java/org/signal/pagingtest/MainActivity.java
package org.signal.pagingtest;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.signal.paging.PagingController;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public class MainActivity extends AppCompatActivity implements EventListener {

  private MainViewModel viewModel;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    MyAdapter           adapter       = new MyAdapter(this);
    RecyclerView        list          = findViewById(R.id.list);
    LinearLayoutManager layoutManager = new LinearLayoutManager(this);

    list.setAdapter(adapter);
    list.setLayoutManager(layoutManager);

    viewModel = new ViewModelProvider(this, new ViewModelProvider.NewInstanceFactory()).get(MainViewModel.class);
    adapter.setPagingController(viewModel.getPagingController());
    viewModel.getList().observe(this, newList -> {
      adapter.submitList(newList);
    });

    findViewById(R.id.invalidate_btn).setOnClickListener(v -> {
      viewModel.getPagingController().onDataInvalidated();
    });

    findViewById(R.id.down250_btn).setOnClickListener(v -> {
      int target = Math.min(adapter.getItemCount() - 1, layoutManager.findFirstVisibleItemPosition() + 250);
      layoutManager.scrollToPosition(target);
    });

    findViewById(R.id.up250_btn).setOnClickListener(v -> {
      int target = Math.max(0, layoutManager.findFirstVisibleItemPosition() - 250);
      layoutManager.scrollToPosition(target);
    });

    findViewById(R.id.prepend_btn).setOnClickListener(v -> {
      viewModel.prependItems();
    });
  }

  @Override
  public void onItemClicked(String key) {
    viewModel.onItemClicked(key);
  }

  static class MyAdapter extends RecyclerView.Adapter<MyViewHolder> {

    private final static int TYPE_NORMAL      = 1;
    private final static int TYPE_PLACEHOLDER = -1;

    private final EventListener listener;
    private final List<Item>    data;

    private PagingController<String> controller;

    public MyAdapter(@NonNull EventListener listener) {
      this.listener = listener;
      this.data     = new ArrayList<>();

      setHasStableIds(true);
    }

    @Override
    public int getItemViewType(int position) {
      return getItem(position) == null ? TYPE_PLACEHOLDER : TYPE_NORMAL;
    }

    @Override
    public int getItemCount() {
      return data.size();
    }

    @Override
    public long getItemId(int position) {
      Item item = getItem(position);
      if (item != null) {
        return item.key.hashCode();
      } else {
        return 0;
      }
    }

    @Override
    public @NonNull MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
      switch (viewType) {
        case TYPE_NORMAL:
        case TYPE_PLACEHOLDER:
          return new MyViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item, parent, false));
        default:
          throw new AssertionError();
      }
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
      holder.bind(getItem(position), position, listener);
    }

    private Item getItem(int index) {
      if (controller != null) {
        controller.onDataNeededAroundIndex(index);
      }
      return data.get(index);
    }

    void setPagingController(PagingController<String> pagingController) {
      this.controller = pagingController;
    }

    void submitList(List<Item> list) {
      DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {
        @Override
        public int getOldListSize() {
          return data.size();
        }

        @Override
        public int getNewListSize() {
          return list.size();
        }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
          String oldKey = Optional.ofNullable(data.get(oldItemPosition)).map(item -> item.key).orElse(null);
          String newKey = Optional.ofNullable(list.get(newItemPosition)).map(item -> item.key).orElse(null);

          return Objects.equals(oldKey, newKey);
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
          Long oldKey = Optional.ofNullable(data.get(oldItemPosition)).map(item -> item.timestamp).orElse(null);
          Long newKey = Optional.ofNullable(list.get(newItemPosition)).map(item -> item.timestamp).orElse(null);

          return Objects.equals(oldKey, newKey);
        }
      }, false);

      result.dispatchUpdatesTo(this);

      data.clear();
      data.addAll(list);
    }
  }

  static class MyViewHolder extends RecyclerView.ViewHolder {

    TextView textView;

    public MyViewHolder(@NonNull View itemView) {
      super(itemView);
      textView = itemView.findViewById(R.id.text);
    }

    void bind(@Nullable Item item, int position, @NonNull EventListener listener) {
      if (item != null) {
        textView.setText(position + " | " + item.key.substring(0, 13) + " | " + System.currentTimeMillis());
        textView.setOnClickListener(v -> {
          listener.onItemClicked(item.key);
        });
      } else {
        textView.setText(position + " | PLACEHOLDER");
        textView.setOnClickListener(null);
      }
    }
  }
}

File: glide-webp/app/src/main/java/org/signal/glide/webp/app/MainActivity.kt
/*
 * Copyright 2023 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.signal.glide.webp.app

import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.bumptech.glide.Glide
import com.bumptech.glide.GlideBuilder
import com.bumptech.glide.load.engine.DiskCacheStrategy
import org.signal.core.util.dp

/**
 * Main activity for this sample app.
 */
class MainActivity : AppCompatActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.main_activitiy)

    Glide.init(
      this,
      GlideBuilder()
        .setLogLevel(Log.VERBOSE)
    )

    val context = this

    findViewById<RecyclerView>(R.id.list).apply {
      adapter = ImageAdapter()
      layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
    }
  }

  class ImageAdapter : RecyclerView.Adapter<ImageViewHolder>() {

    private val data: List<String> = listOf(
      "test_01.webp",
      "test_02.webp",
      "test_03.webp",
      "test_04.webp",
      "test_05.webp",
      "test_06_lossless.webp",
      "test_06_lossy.webp",
      "test_07_lossless.webp",
      "test_09_large.webp"
    )

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ImageViewHolder {
      return ImageViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.image_item, parent, false))
    }

    override fun getItemCount(): Int {
      return data.size
    }

    override fun onBindViewHolder(holder: ImageViewHolder, position: Int) {
      holder.bind(data[position])
    }
  }

  class ImageViewHolder(itemView: View) : ViewHolder(itemView) {
    val image: ImageView by lazy { itemView.findViewById(R.id.image) }

    fun bind(filename: String) {
      Glide.with(itemView)
        .load(Uri.parse("file:///android_asset/$filename"))
        .skipMemoryCache(true)
        .diskCacheStrategy(DiskCacheStrategy.NONE)
        .override(250.dp)
        .into(image)
    }
  }
}
